### ISSUE TITLE
Git doesn't detect renames after 200 files

### ISSUE THREAD

**Issue by timotei** [git-for-windows/git#1098]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.12.0.windows.1
built from commit: d4baef1cdfef9de62d43f9682a98ba57adc842e4
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.14393]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt


Path Option: Cmd
Plink Path: E:\bin\putty\PLINK.EXE
SSH Option: Plink
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Nope

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
I posted here a reproduction repo: https://github.com/timotei/git-rename-limit-testcase It has the reproduction path in the README.
```

 - What did you expect to occur after running these commands?

All files are detected as renamed, since they have similarity index 95%

 - What actually happened instead?

A rename is not detected anymore after the 201th file

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

https://github.com/timotei/git-rename-limit-testcase

**Comment by PhilipOakley** [comment]
> A rename is not detected anymore after the 201th file

Do you know a recent version where it was working for you?

Other ideas..
Are you able to bisect the issue with the SDK?
Is it mentioned in the main git mailing list release notes https://github.com/git/git/tree/master/Documentation/RelNotes ?

**Comment by kostix** [comment]
I may be completely out of touch with the issue but does setting `diff.renameLimit` to a high value (like 2000 or something) help?

> `diff.renameLimit`
>
> The number of files to consider when performing the copy/rename detection;
> equivalent to the `git diff` option `-l`.

I once needed it when dealing with a subtree merge of an insanely huge codebase featuring long pathames (PHP).

**Comment by timotei** [comment]
@PhilipOakley Well, I don't know if other version works.

I just tested on a Linux machine with git 2.1.4 and it seems I've got the same issue. Maybe I should start a discussion (also) on the git mailing list.

@kostix Seems to reproduce even with that value set to something like 3000

**Comment by rimrul** [comment]
This does not seem to reproduce anymore with Git for Windows 2.26.2

**Comment by derrickstolee** [comment]
> This does not seem to reproduce anymore with Git for Windows 2.26.2

This is likely due to recent directory-based rename detection created by @newren.

**Comment by newren** [comment]
Actually, it sounds like the bugs that were fixed by the en/rename-progress topic merged in late 2017 and released as part of Git 2.16.0.  (I'm slightly surprised that the number someone ran into was 201 for a symmetric case; when it's symmetric, I would have expected the breaking point to appear closer to sqrt(32768), but things did get a little funny in diffcore-rename).

en/rename-progress pre-dated the directory rename detection work by a little bit, though was the same general timeframe.  But yeah, "rename detection has arbitrary and anemic small limits separate from and often smaller than diff.renameLimit -- with often no warning provided and never any control to override" was fixed years ago.

**Comment by dscho** [comment]
Excellent! Thank you all!
